diff --git a/hid-uclogic-params.c b/hid-uclogic-params.c
index 86ea181..557cd85 100644
--- a/hid-uclogic-params.c
+++ b/hid-uclogic-params.c
@@ -1221,9 +1221,11 @@ int uclogic_params_init(struct uclogic_params *params,
 	/* The resulting parameters (noop) */
 	struct uclogic_params p = {0, };
 
-	/* Check arguments */
+        /* Check arguments */
 	if (params == NULL || hdev == NULL
-#if KERNEL_VERSION(4, 14, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 3, 0) <= LINUX_VERSION_CODE
+	    || !hid_is_usb(hdev)
+#elif KERNEL_VERSION(4, 14, 0) <= LINUX_VERSION_CODE
 	    || !hid_is_using_ll_driver(hdev, &usb_hid_driver)
 #endif
 	   ) {
