diff --git a/setup.py b/setup.py
index c3782dc..e11be44 100644
--- a/setup.py
+++ b/setup.py
@@ -37,7 +37,7 @@ setup(name='obs_service_gbs',
       url=tag_from_spec('URL'),
       license=tag_from_spec('License'),
       packages=['obs_service_gbs'],
-      data_files=[('/usr/lib/obs/service', ['service/gbs',
-                                            'service/gbs.service']),
+      data_files=[('/usr/libexec/obs/service', ['service/gbs',
+                                                'service/gbs.service']),
                   ('/etc/obs/services', ['config/gbs'])],
      )
