--- CMakeLists.txt	2013-07-30 09:28:51.507189646 +0400
+++ CMakeLists.txt.fixed	2013-07-30 09:28:45.813189352 +0400
@@ -184,7 +184,7 @@
 
 MESSAGE(STATUS "Looking modules in ${CMAKE_MODULE_PATH}")
 
-set ( CMAKE_C_FLAGS     "${CMAKE_C_FLAGS} -Werror -Wall" )
+set ( CMAKE_C_FLAGS     "${CMAKE_C_FLAGS} -Wall" )
 set ( CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS} -g -O3" )
 set ( CMAKE_C_FLAGS_DEBUG     "${CMAKE_C_FLAGS} -g3 -O0" )
 
