--- a/setup.py	2021-03-03 11:24:40.393509918 +0300
+++ b/setup.py	2021-03-03 11:24:42.568510112 +0300
@@ -27,7 +27,7 @@
     version = "0.0"
 
     try:
-        popen = subprocess.Popen('dpkg-parsechangelog', stdout=subprocess.PIPE)
+        popen = subprocess.Popen('dpkg-parsechangelog', stdout=subprocess.PIPE, encoding='utf8')
         out, ret = popen.communicate()
         for line in out.split('\n'):
             if line.startswith('Version:'):
@@ -31,12 +31,12 @@
         out, ret = popen.communicate()
         for line in out.split('\n'):
             if line.startswith('Version:'):
-                version = line.split(' ')[1].strip()
+                version = line.split(' ')[1].strip().split('-', 1)[0]
                 break
     except OSError:
         pass # Failing is fine, we just can't print the version then
 
     with open('gbp/version.py', 'w') as f:
         f.write('"The current gbp version number"\n')
-        f.write('gbp_version="%s"\n' % version)
+        f.write('gbp_version="%s"\n' % version.split('-', 1)[0])
 
     return version